File: solutions/csharp/split-second-stopwatch/1/SplitSecondStopwatch.cs
#pragma warning disable IDE0079
#pragma warning disable IDE0130
#pragma warning disable CA1050

public class SplitSecondStopwatch(TimeProvider time)
{
    private long _startTimestamp;
    private long _currentLapStartTimestamp;
    private TimeSpan _currentLapElapsed;
    private TimeSpan _totalElapsed;

    public StopwatchState State { get; private set; } = StopwatchState.Ready;

    public TimeSpan CurrentLap => State == StopwatchState.Running
        ? _currentLapElapsed + time.GetElapsedTime(_currentLapStartTimestamp)
        : _currentLapElapsed;

    public TimeSpan Total => State == StopwatchState.Running
        ? _totalElapsed + time.GetElapsedTime(_startTimestamp)
        : _totalElapsed;

    public IReadOnlyCollection<TimeSpan> PreviousLaps { get; private set; } = [];

    public void Start()
    {
        if (State == StopwatchState.Running)
        {
            throw new InvalidOperationException("Start can only be called when the stopwatch is not running.");
        }

        long currentTimestamp = time.GetTimestamp();

        switch (State)
        {
            case StopwatchState.Ready:
                _startTimestamp = currentTimestamp;
                _currentLapStartTimestamp = currentTimestamp;
                _currentLapElapsed = TimeSpan.Zero;
                _totalElapsed = TimeSpan.Zero;
                break;
            case StopwatchState.Stopped:
                _startTimestamp = currentTimestamp;
                _currentLapStartTimestamp = currentTimestamp;
                break;
        }

        State = StopwatchState.Running;
    }

    public void Stop()
    {
        if (State != StopwatchState.Running)
        {
            throw new InvalidOperationException("Stop can only be called when the stopwatch is running.");
        }

        _currentLapElapsed += time.GetElapsedTime(_currentLapStartTimestamp);
        _totalElapsed += time.GetElapsedTime(_startTimestamp);

        State = StopwatchState.Stopped;
    }

    public void Reset()
    {
        if (State == StopwatchState.Ready || State == StopwatchState.Running)
        {
            throw new InvalidOperationException("Reset can only be called when the stopwatch is stopped.");
        }

        State = StopwatchState.Ready;
        _startTimestamp = 0;
        _currentLapStartTimestamp = 0;
        _currentLapElapsed = TimeSpan.Zero;
        _totalElapsed = TimeSpan.Zero;
        PreviousLaps = [];
    }

    public void Lap()
    {
        if (State != StopwatchState.Running)
        {
            throw new InvalidOperationException("Lap can only be called when the stopwatch is running.");
        }

        TimeSpan currentLapTime = _currentLapElapsed + time.GetElapsedTime(_currentLapStartTimestamp);

        List<TimeSpan> previousLapsList = PreviousLaps.ToList();
        previousLapsList.Add(currentLapTime);
        PreviousLaps = previousLapsList.AsReadOnly();

        _currentLapElapsed = TimeSpan.Zero;
        _currentLapStartTimestamp = time.GetTimestamp();
    }
}

#pragma warning disable IDE0079
#pragma warning disable IDE0130
#pragma warning disable CA1050
public enum StopwatchState
{
    Ready,
    Running,
    Stopped
}

File: solutions/java/split-second-stopwatch/1/src/main/java/SplitSecondStopwatch.java
import java.util.ArrayList;
import java.util.List;

public class SplitSecondStopwatch {
    private final SplitSecondStopwatchHelpers stopwatchHelpers;
    private SplitSecondStopwatch(SplitSecondStopwatchHelpers stopwatchHelpers) {
        this.stopwatchHelpers = stopwatchHelpers;
    }

    public SplitSecondStopwatch(){
        this(new SplitSecondStopwatchHelpers());
    }

    private StopwatchState currentState = StopwatchState.READY;
    private int currentLapSeconds = 0;
    private int totalSeconds = 0;
    private final List<Integer> previousLapSecondsList = new ArrayList<>();

    public void start() {
        switch (this.currentState) {
            case READY, STOPPED
                    -> this.currentState = StopwatchState.RUNNING;

            case RUNNING
                    -> throw new IllegalStateException("cannot start an already running stopwatch");
        }
    }

    public void stop() {
        if (this.currentState != StopwatchState.RUNNING) {
            throw new IllegalStateException("cannot stop a stopwatch that is not running");
        }

        this.currentState = StopwatchState.STOPPED;
    }

    public void reset() {
        if (this.currentState != StopwatchState.STOPPED) {
            throw new IllegalStateException("cannot reset a stopwatch that is not stopped");
        }

        this.currentState = StopwatchState.READY;
        this.currentLapSeconds = 0;
        this.totalSeconds = 0;
        this.previousLapSecondsList.clear();
    }

    public void lap() {
        if (this.currentState != StopwatchState.RUNNING) {
            throw new IllegalStateException("cannot lap a stopwatch that is not running");
        }

        this.previousLapSecondsList.add(this.currentLapSeconds);
        this.currentLapSeconds = 0;
    }

    public String state() {
        return switch (this.currentState) {
            case READY -> "ready";
            case RUNNING -> "running";
            case STOPPED -> "stopped";
        };
    }

    public String currentLap() {
        return this.stopwatchHelpers.formatSeconds(this.currentLapSeconds);
    }

    public String total() {
        return this.stopwatchHelpers.formatSeconds(this.totalSeconds);
    }

    public List<String> previousLaps() {
        return this.previousLapSecondsList.stream()
                       .map(this.stopwatchHelpers::formatSeconds)
                       .toList();
    }

    public void advanceTime(String timeString) {
        if (this.currentState == StopwatchState.RUNNING) {
            int secondsToAdd = this.stopwatchHelpers.parseTimeStringToSeconds(timeString);
            this.currentLapSeconds += secondsToAdd;
            this.totalSeconds += secondsToAdd;
        }
    }
}
